src/libs/OptionsListUtils.js

```diff
@@ -736,6 +736,132 @@ function getCategoryListSections(categories, recentlyUsedCategories, selectedOpt
     return categorySections;
 }
 
+/**
+ * Transforms the provided tags into objects with a specific structure.
+ *
+ * @param {Object[]} tags - an initial tag array
+ * @param {Boolean} tags[].enabled - a flag to enable/disable option in a list
+ * @param {String} tags[].name - a name of an option
+ * @returns {Array<Object>}
+ */
+function getTagsOptions(tags) {
+    const options = [];
+
+    _.each(tags, (tag) =>
+        options.push({
+            text: tag.name,
+            keyForList: tag.name,
+            searchText: tag.name,
+            tooltipText: tag.name,
+            isDisabled: !tag.enabled,
+        }),
+    );
+
+    return options;
+}
+
+/**
+ * Build the section list for tags
+ *
+ * @param {Object[]} tags
+ * @param {String} tags[].name
+ * @param {Boolean} tags[].enabled
+ * @param {String[]} recentlyUsedTags
+ * @param {Object[]} selectedOptions
+ * @param {String} selectedOptions[].name
+ * @param {String} searchInputValue
+ * @param {Number} maxRecentReportsToShow
+ * @returns {Array<Object>}
+ */
+function getTagListSections(tags, recentlyUsedTags, selectedOptions, searchInputValue, maxRecentReportsToShow) {
+    const tagSections = [];
+    const numberOfTags = _.size(tags);
+    let indexOffset = 0;
+
+    if (!_.isEmpty(searchInputValue)) {
+        const searchTags = _.filter(tags, (tag) => tag.name.toLowerCase().includes(searchInputValue.toLowerCase()));
+
+        tagSections.push({
+            // "Search" section
+            title: '',
+            shouldShow: false,
+            indexOffset,
+            data: getTagsOptions(searchTags),
+        });
+
+        return tagSections;
+    }
+
+    if (numberOfTags < CONST.TAG_LIST_THRESHOLD) {
+        tagSections.push({
+            // "All" section when items amount less than the threshold
+            title: '',
+            shouldShow: false,
+            indexOffset,
+            data: getTagsOptions(tags),
+        });
+
+        return tagSections;
+    }
+
+    const selectedOptionNames = _.map(selectedOptions, (selectedOption) => selectedOption.name);
+    const filteredRecentlyUsedTags = _.map(
+        _.filter(recentlyUsedTags, (tag) => !_.includes(selectedOptionNames, tag)),
+        (tag) => {
+            const tagObject = _.find(tags, (t) => t.name === tag);
+            return {
+                name: tag,
+                enabled: tagObject && tagObject.enabled,
```

Comment:
AFAIK, `tagObject` is not a boolean type, we should handle it.

```suggestion
                enabled: Boolean(tagObject && tagObject.enabled),
```